File: src/main/scala/search/mcts/montecarlo/MonteCarloTreeSearchTest.scala
package search.mcts.montecarlo

import org.junit.jupiter.api.Assertions._
import org.junit.jupiter.api.{Nested, RepeatedTest, Test}
import search.mcts.tictactoe._
import search.mcts.tree.Node

import scala.collection.mutable.ArrayBuffer

object MonteCarloTreeSearchTest {

  @Nested
  class ticTacToeTest {

    @RepeatedTest(100)
    def fullTest(): Unit = {
      var board: Board = TicTacToeBoard()
      var player       = 1
      var round        = 1
      while (board.checkStatus == InProgress) {
        println(s"${">" * 25} ROUND $round ${"<" * 25}")
        round += 1
        board = MonteCarloTreeSearch.findNextMove(board, player)
        player = State.getOpponent(player)
      }
      val winStatus    = board.checkStatus
      assertEquals(Draw, winStatus)
    }

    @Test
    def findNextMove(): Unit = {
      val board  = TicTacToeBoard()
      val player = 1
      MonteCarloTreeSearch.findNextMove(board, player)
    }

    @Test
    def testUCT(): Unit = {
      val uctValue = 15.79
      assertEquals(UCT.uctValue(600, 300, 20), uctValue, 0.01)
    }

    @Test
    def performMoveEmptyBoard(): Unit = {
      var board                  = TicTacToeBoard()
      val initAvailablePositions = board.emptyPositions.size
      board = board.performMove(P1.status, Position(1, 1))
      val availablePositions     = board.emptyPositions.size
      assertEquals(8, availablePositions)
      assertTrue(initAvailablePositions > availablePositions)
    }

    @Test
    def getAllPossibleStates(): Unit = {
      val initState      = State(2, TicTacToeBoard())
      val possibleStates = initState.getAllPossibleStates
      assertEquals(9, possibleStates.length)
      assertTrue(possibleStates.forall(_.playerNo == 1))
      assertTrue(possibleStates.forall(_.board.asInstanceOf[TicTacToeBoard].boardValues.flatten.contains(1)))
    }

    @Test
    def expandNode(): Unit = {
      val emptyBoard = TicTacToeBoard()
      val rootNode   = Node(State(1, emptyBoard), None, ArrayBuffer.empty[Node])
      MonteCarloTreeSearch.expandNode(rootNode)
      assertEquals(9, rootNode.childArray.length)
      assertTrue(rootNode.childArray.forall(_.parent.contains(rootNode)))
      assertTrue(rootNode.childArray.forall(_.state.playerNo != rootNode.state.playerNo))
    }

    @Test
    def backPropagateRoot(): Unit = {
      val emptyBoard = TicTacToeBoard()
      val rootNode   = Node(State(1, emptyBoard), None, ArrayBuffer.empty[Node])
      assertEquals(0, rootNode.state.visitCount)
      MonteCarloTreeSearch.backPropagation(rootNode, 1)
      assertEquals(1, rootNode.state.visitCount)
    }

    @Test
    def simulateRandomPlayout(): Unit = {
      val emptyBoard = TicTacToeBoard()
      val rootNode   = Node(State(2, emptyBoard), None, ArrayBuffer.empty[Node])
      val node       = MonteCarloTreeSearch.selectPromisingNode(rootNode)
      MonteCarloTreeSearch.expandNode(node)
      println(node)
      println(MonteCarloTreeSearch.simulateRandomPlayout(node, 2))
    }

    @Test
    def backPropagateChildNode(): Unit = {
      val emptyBoard = TicTacToeBoard()
      val rootNode   = Node(State(2, emptyBoard), None, ArrayBuffer.empty[Node])

      MonteCarloTreeSearch.expandNode(rootNode)
      val playoutResult = MonteCarloTreeSearch.simulateRandomPlayout(rootNode.childArray.head, 2)
      MonteCarloTreeSearch.backPropagation(rootNode.childArray.head, playoutResult.status)
      assertEquals(1, rootNode.state.visitCount)

      MonteCarloTreeSearch.expandNode(rootNode.childArray.head)
      val playoutResult2 = MonteCarloTreeSearch.simulateRandomPlayout(rootNode.childArray.head.childArray.head, 1)
      MonteCarloTreeSearch.backPropagation(rootNode.childArray.head.childArray.head, playoutResult2.status)
      assertEquals(2, rootNode.state.visitCount)
    }

    @Test
    def checkStatus(): Unit = {
      val boardDraw       = TicTacToeBoard(3, Vector(Vector(1, 2, 1), Vector(1, 2, 2), Vector(2, 1, 1)))
      assertEquals(Draw, boardDraw.checkStatus)
      val boardP1Win      = TicTacToeBoard(3, Vector(Vector(1, 2, 1), Vector(1, 1, 2), Vector(2, 2, 1)))
      assertEquals(P1, boardP1Win.checkStatus)
      val boardP2Win      = TicTacToeBoard(3, Vector(Vector(1, 1, 2), Vector(1, 2, 1), Vector(2, 1, 2)))
      assertEquals(P2, boardP2Win.checkStatus)
      val boardInProgress = TicTacToeBoard(3, Vector(Vector(0, 0, 1), Vector(0, 2, 2), Vector(1, 0, 0)))
      assertEquals(InProgress, boardInProgress.checkStatus)
      val P1WinExtra      = TicTacToeBoard(3, Vector(Vector(2, 1, 0), Vector(2, 2, 0), Vector(1, 1, 1)))
      assertEquals(P1, P1WinExtra.checkStatus)
    }

    @Test
    def setScore(): Unit = {
      val node = Node(State(1, TicTacToeBoard()), None, ArrayBuffer.empty[Node])
      node.addScore(10)
      assertEquals(10, node.state.winScore)
      node.setScore()
      assertEquals(Int.MinValue, node.state.winScore)
    }

    @Test
    def addMinScoreToParent(): Unit = {
      val node = Node(State(1, TicTacToeBoard()), None, ArrayBuffer.empty[Node])
      MonteCarloTreeSearch.expandNode(node)
      node.childArray.head.parent.get.addScore(Int.MinValue)
      println(node)
    }

    @RepeatedTest(10)
    def findNextStepCustomStart1(): Unit = {
      val board    = TicTacToeBoard(
        3,
        Vector(
          Vector(2, 1, 0),
          Vector(0, 2, 0),
          Vector(1, 0, 1)
        )
      )
      val newBoard = MonteCarloTreeSearch.findNextMove(board, 2)
      assertEquals(2, newBoard.asInstanceOf[TicTacToeBoard].boardValues(2)(1))
    }
  }
}

File: src/main/java/mcts/montecarlo/MonteCarloTreeSearchTest.java
package mcts.montecarlo;

import mcts.tictactoe.Board;

import mcts.tictactoe.Position;
import mcts.tree.Node;
import mcts.tree.Tree;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class MonteCarloTreeSearchTest {
    private Tree gameTree;
    private MonteCarloTreeSearch mcts;


    @BeforeEach
    void setUp() {
        gameTree = new Tree();
        mcts = new MonteCarloTreeSearch();
    }

    @AfterEach
    void tearDown() {
    }

    @RepeatedTest(100)
    public void givenEmptyBoard_whenSimulateInterAIPlay_thenGameDraw() {
        int boardSize = 3;
        Board board = new Board(boardSize);

        int player = Board.P1;
        int totalMoves = boardSize * boardSize;
        for (int i = 0; i < totalMoves; i++) {
            board = mcts.findNextMove(board, player);
            if (board.checkStatus() != -1) {
                break;
            }
            player = 3 - player;
        }
        int winStatus = board.checkStatus();
        assertEquals(winStatus, Board.DRAW);
    }

    @Test
    public void givenStats_whenGetUCTForNode_thenUCTMatchesWithManualData() {
        double uctValue = 15.79;
        assertEquals(UCT.uctValue(600, 300, 20), uctValue, 0.01);
    }

    @Test
    public void giveninitBoardState_whenGetAllPossibleStates_thenNonEmptyList() {
        State initState = gameTree.getRoot().getState();
        List<State> possibleStates = initState.getAllPossibleStates();
        assertTrue(possibleStates.size() > 0);
    }

    @Test
    public void givenEmptyBoard_whenPerformMove_thenLessAvailablePossitions() {
        Board board = new Board();
        int initAvailablePositions = board.getEmptyPositions().size();
        board.performMove(Board.P1, new Position(1, 1));
        int availablePositions = board.getEmptyPositions().size();
        assertTrue(initAvailablePositions > availablePositions);
    }

    @Test
    public void givenEmptyBoard_whenLevel1VsLevel3_thenLevel3WinsOrDraw() {
        Board board = new Board();
        MonteCarloTreeSearch mcts1 = new MonteCarloTreeSearch();
        mcts1.setLevel(1);
        MonteCarloTreeSearch mcts3 = new MonteCarloTreeSearch();
        mcts3.setLevel(3);

        int player = Board.P1;
        int totalMoves = Board.DEFAULT_BOARD_SIZE * Board.DEFAULT_BOARD_SIZE;
        for (int i = 0; i < totalMoves; i++) {
            if (player == Board.P1)
                board = mcts3.findNextMove(board, player);
            else
                board = mcts1.findNextMove(board, player);

            if (board.checkStatus() != -1) {
                break;
            }
            player = 3 - player;
        }
        int winStatus = board.checkStatus();
        assertTrue(winStatus == Board.DRAW || winStatus == Board.P1);
    }

    @Test
    public void addMinScore() {
        Node node = new Node();
        node.getState().addScore(10);
        assertEquals(10, node.getState().getWinScore());
        node.getState().setWinScore(Integer.MIN_VALUE);
        assertEquals(Integer.MIN_VALUE, node.getState().getWinScore());
    }

    @RepeatedTest(10)
    public void nextStepCustom() {
        int[][] ar = {{2, 1, 0},{0, 2, 0},{1, 0, 1}};
        Board board = new Board(ar);
        Board newBoard = mcts.findNextMove(board, 2);
        assertEquals(2, newBoard.getBoardValues()[2][1]);
    }

    @RepeatedTest(10)
    public void simulateRandomPlayout() {
        int[][] ar = {{2, 1, 0},{0, 2, 0},{1, 0, 1}};
        Board board = new Board(ar);
        State state = new State(board);
        state.setPlayerNo(1);
        Node rootNode = new Node(state);
        Node node = mcts.selectPromisingNode(rootNode);
        mcts.expandNode(node);
        Node nodeToExplore = node.getRandomChildNode();
        int r = mcts.simulateRandomPlayout(nodeToExplore);
        mcts.backPropogation(nodeToExplore, r);
        System.out.println(node);
    }

}
